flocker/control/httpapi.py

```diff
@@ -212,6 +213,112 @@ def saved(ignored):
         saving.addCallback(saved)
         return saving
 
+    @app.route("/configuration/datasets/<dataset_id>", methods=['POST'])
+    @user_documentation(
+        """
+        Update an existing dataset.
+        """,
+        examples=[
+            u"update dataset with primary",
+            u"update dataset with unknown dataset id",
+        ]
+    )
+    @structured(
+        inputSchema={'$ref': '/v1/endpoints.json#/definitions/datasets'},
+        outputSchema={'$ref': '/v1/endpoints.json#/definitions/datasets'},
+        schema_store=SCHEMAS
+    )
+    def update_dataset(self, dataset_id, primary=None):
+        """
+        Update an existing dataset in the cluster configuration.
+
+        :param unicode dataset_id: The unique identifier of the dataset.  This
+            is a string giving a UUID (per RFC 4122).
+
+        :param unicode primary: The address of the node to which the dataset
+            will be moved.
+
+        :return: A ``dict`` describing the dataset which has been added to the
+            cluster configuration or giving error information if this is not
+            possible.
+        """
+        # Get the current configuration.
+        deployment = self.persistence_service.get()
+
+        # Find all manifestations of requested dataset.
+        manifestations_and_nodes = list(
+            other_manifestations_from_deployment(deployment, dataset_id)
+        )
+
+        # There are no manifestations containing the requested dataset.
+        if not manifestations_and_nodes:
+            raise DATASET_NOT_FOUND
+
+        # Lookup the node that has a primary Manifestation (if any)
+        primary_manifestation, origin_node = [
+            (manifestation, node)
+            for manifestation, node in manifestations_and_nodes
+            if manifestation.primary
+        ][0]
```

Comment:
If you're going to have it blow up if no primary manifestation exists then please file followup issue (with link from here) to support that use case, which we will have with Cinder and EBS.